Update testing documentation with testing guidelines. (#87928)
llvm/llvm-project: libcxx/docs/TestingLibcxx.rst

  ``libcxx/test/libcxx``. The structure of this directories follows the
  structure of ``libcxx/test/std``.

Principles of testing
---------------------

Tests are a practical way to validate the correctness of the code. As such, they contain pragmatic trade offs between
the cost of writing and maintaining the tests and the value they provide. Please consider the following principles when
writing tests:

- **Consider the next reader**

>     Tests should be obvious to the future reader. Avoid too much boiler plate or other

Reviewer: We should avoid terms like "obvious", as that's a subjective metric. Things that are obvious to you might not be obvious to me, and vice versa.
```suggestion
    Unit tests should be written so that anyone who lacks context can immediately identify that they are correct. Avoid too much boiler plate or other
```

    distractions. Ensure each test has enough context to understand what it is testing. Avoid gratuitous use of advanced
    test features or abstractions.

> - **Consider the effect of time**

Author: I think its' clear enough. Sorry I wish I had a better response.

    Tests should be resilient to the effects of time. Tests are not static; They are
    living documents that change. As they change the original intent of a test can become less clear.

- **Consider the edge cases carefully**

    Undefined behavior and edge cases are often the source of bugs. Tests should exercise these cases to ensure that

>     the code under test behaves correctly. It's important to write tests for the easy cases as well as the hard ones.

Reviewer 2: I don't think we should do that. I see TDD as a tool, sometimes it's the right tool for the job sometimes not.

  ``libcxx/test/libcxx``. The structure of this directories follows the
  structure of ``libcxx/test/std``.

> Principles of testing

Reviewer 3: I think I might move this before the "structure of the testing related directories" since it's more general?

Author: Done.

---------------------

Tests are a practical way to validate the correctness of the code. As such, they contain pragmatic trade offs between
the cost of writing and maintaining the tests and the value they provide. Please consider the following principles when
writing tests:

> - **Consider the next reader**

Reviewer 3: Per our live discussion just now, it might make sense to turn this over and first state the values, and then provide a list of recommended practices that derive from those values. For example:

```
1. Consider the next reader
2. Consider the effect of time
3. etc..

From this, we derive the following set of recommended practices. These practices are not hard rules, they are just recommendations that often make sense:
- Write plain text comments explaining what your test is testing when it's not obvious. For example, when testing multiple edge cases for a function, it's often a good idea to document [bla bla bla]. This relates to considerations (1) and (2).
- Write focused tests that test a single thing. [bla bla bla]. This relates to consideration (N).
- etc..
```

Reviewer: Do we want people to "consider X" when writing tests, or do we want them to take bold action instead?

As an example, perhaps "tests should be resilient to change over time" gets the developer to consider the effect of time, but also gives something actionable in the title (which is probably what will be remembered).

Author: These are meant as starting points for a conversation, or base axioms.

They're not meant to be calls to action, but rather concerns we all share that allow us to discuss the utility of one particular action over another.

The idea is that we can have better, more meaningful disagreements if we all start with the same base concerns.

>     Tests should be obvious to tho future reader. Avoid too much boiler plate or other

Reviewer 4: ```suggestion
    Tests should be obvious to the future reader. Avoid too much boiler plate or other
```

    distractions. Ensure each test has enough context to understand what it is testing. Avoid gratuitous use of advanced

>     test features or abstractions.

Reviewer 3: Add: We also strongly encourage the intent of a test and the reason for its existence to be documented in plain text comments.

This would address Mark's comment below.

- **Consider the effect of time**

    Tests should be resilient to the effects of time. Tests are not static; They are
    living documents that change over time. The intent of a test can become less clear over time.

- **Consider the edge cases carefully**

    Undefined behavior and edge cases are often the source of bugs. Tests should exercise these cases to ensure that
    the code under test behaves correctly. It's important to write tests for the easy cases as well as the hard ones.

- **Consider the focus**

    Each test case should test a single concern. Ideally a test should only fail when this concern

>     is violated. Focused tests are not flakey. If a test case covers multiple concerns, consider splitting it into multiple

Reviewer 3: ```suggestion
    is violated. Focused tests are not flaky. If a test case covers multiple concerns, consider splitting it into multiple
```

I think?

    test cases.

> 

Reviewer 2: Since we (Louis, you, and I) discussed adding more comments to test I'd like to see that here too.

> 

Reviewer 2: Another thing I'd like to see is the consideration when a fuzz test should be added. 

I also like to see a (short) section on fuzz testing in this document. Until Nikolas started to improve them I wasn't aware we had them at all; and until this week I wasn't aware they are tested in OSS-Fuzz.

Note that these are principles, not rules. There are cases where it is appropriate to break these principles.
During code review, if you feel that a test is not following these principles, please discuss it with the author.

Remember code review is a **consensus building** process, not a **gatekeeping** process. Your fellow engineers are

> all working towards the same goal: to make the code better.

Reviewer 3: While I agree with this and there is value in saying it somewhere, I don't think this belongs in this section. This section is about testing considerations, while these sentences would belong in a section discussing general code review principles, if/once we add one.
